src/software/sensor_fusion/filter/ball_filter.cpp

```diff
@@ -329,49 +313,85 @@ std::optional<TimestampedBallState> BallFilter::estimateBallState(
         auto velocity_direction_along_regression_line =
             velocity_estimate->average_velocity.project(
                 regression_line.toNormalUnitVector().perpendicular());
-        Vector filtered_velocity = velocity_direction_along_regression_line.normalize(
+        filtered_velocity = velocity_direction_along_regression_line.normalize(
             velocity_estimate->average_velocity_magnitude);
-
-        BallState ball_state(filtered_ball_position, filtered_velocity,
-                             latest_ball_detection.distance_from_ground);
-        return TimestampedBallState(ball_state, latest_ball_detection.timestamp);
     }
+
+    return filtered_velocity;
 }
 
-std::optional<TimestampedBallState> BallFilter::getFilteredData(
-    const std::vector<BallDetection> &new_ball_detections, const Field &field)
+std::optional<BallFilter::BallVelocityEstimate> BallFilter::estimateBallVelocity(
+    boost::circular_buffer<BallDetection> ball_detections,
+    const std::optional<Line> &ball_regression_line)
 {
-    addNewDetectionsToBuffer(new_ball_detections, field);
+    // Sort the detections in increasing order before processing. This places the oldest
+    // detections (smallest timestamp) at the front of the buffer, and the most recent
+    // detections (with the largest timestamp) at the end of the buffer
+    std::sort(ball_detections.begin(), ball_detections.end());
 
-    if (ball_detection_buffer.size() >= 2)
+    std::vector<Vector> ball_velocities;
+    std::vector<double> ball_velocity_magnitudes;
+    for (unsigned i = 1; i < ball_detections.size(); i++)
     {
-        std::optional<TimestampedBallState> filtered_ball =
-            estimateBallState(ball_detection_buffer);
-        if (filtered_ball)
-        {
-            return *filtered_ball;
-        }
-        else
+        for (unsigned j = i; j < ball_detections.size(); j++)
         {
-            BallState ball_state(ball_detection_buffer.front().position, Vector(0, 0),
-                                 ball_detection_buffer.front().distance_from_ground);
-            TimestampedBallState timestamped_ball_state(
-                ball_state, ball_detection_buffer.front().timestamp);
-            return timestamped_ball_state;
+            BallDetection previous_detection = ball_detections.at(i - 1);
+            BallDetection current_detection  = ball_detections.at(j);
+
+            Duration time_diff =
+                current_detection.timestamp - previous_detection.timestamp;
+            // Avoid division by 0. If we have adjacent detections with the same timestamp
+            // the velocity cannot be calculated
+            if (time_diff.getSeconds() == 0)
```

Comment:
Are there any tests in place for this check?

Reply:
Not explicitly, because the tests are more "black box tests" that don't know about the filter implementation. The tests do implicitly cover the cases where detections could be received "out of order" in time due to the noise added, but there's no guarantee it explicitly tests the case where detections have the same timestamp.